calypso.1
.TH CALYPSO "1" "August 2013" "calypso 1.1" "User Commands"
.SH NAME
calypso \- CalDAV/CardDAV server with git storage
.SH SYNOPSIS
.B calypso
[\fIoptions\fR]
.SH OPTIONS
.TP
\fB\-h\fR, \fB\-\-help\fR
show this help message and exit
.TP
\fB\-v\fR, \fB\-\-version\fR
show version and exit
.TP
\fB\-d\fR, \fB\-\-daemon\fR
launch as daemon
.TP
\fB\-f\fR, \fB\-\-foreground\fR
launch in foreground (opposite of \fB\-\-daemon\fR)
.TP
\fB\-H\fR HOST, \fB\-\-host\fR=\fIHOST\fR
set server hostname
.TP
\fB\-p\fR PORT, \fB\-\-port\fR=\fIPORT\fR
set server port
.TP
\fB\-s\fR, \fB\-\-ssl\fR
use SSL connection
.TP
\fB\-S\fR, \fB\-\-no\-ssl\fR
do not use SSL connection (opposite of \fB\-\-ssl\fR)
.TP
\fB\-k\fR KEY, \fB\-\-key\fR=\fIKEY\fR
private key file
.TP
\fB\-c\fR CERTIFICATE, \fB\-\-certificate\fR=\fICERTIFICATE\fR
certificate file
.HP
\fB\-i\fR IMPORT_DEST, \fB\-\-import\fR=\fIIMPORT_DEST\fR
.TP
\fB\-g\fR, \fB\-\-debug\fR
enable debug logging
.SH AUTHOR
Written by Keith Packard <keithp@keithp.com>
.SH COPYRIGHT
Copyright © 2011 Keith Packard
.br
Copyright © 2008-2011 Guillaume Ayoub
.br
Copyright © 2008 Nicolas Kandel
.br
Copyright © 2008 Pascal Halter
.br
.P
This library is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.
.P
This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.
.P
You should have received a copy of the GNU General Public License
along with Calypso.  If not, see <http://www.gnu.org/licenses/>.


NEWS
==========================
 Calypso - CalDAV Server
==========================

------
 NEWS
------

0.5 - Historical Artifacts
==========================

* Calendar depth
* iPhone support
* MacOS and Windows support
* HEAD requests management
* htpasswd user from calendar path


0.4 - Hot Days Back
===================

* Personal calendars
* HEAD requests
* Last-Modified HTTP header
* ``no-ssl`` and ``foreground`` options
* Default configuration file


0.3 - Dancing Flowers
=====================

* Evolution support
* Version management


0.2 - Snowflakes
================

* Sunbird pre-1.0 support
* SSL connection
* Htpasswd authentication
* Daemon mode
* User configuration
* Twisted dependency removed
* Python 3 support
* Real URLs for PUT and DELETE
* Concurrent modification reported to users
* Many bugs fixed by Roger Wenham


0.1 - Crazy Vegetables
======================

* First release
* Lightning/Sunbird 0.9 compatibility
* Easy installer


calypso.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
#
# This file is part of Calypso - CalDAV/CardDAV/WebDAV Server
# Copyright © 2011 Keith Packard
# Copyright © 2008-2011 Guillaume Ayoub
# Copyright © 2008 Nicolas Kandel
# Copyright © 2008 Pascal Halter
#
# This library is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Calypso.  If not, see <http://www.gnu.org/licenses/>.

# This file is just a script, allow [a-z0-9]* variable names
# pylint: disable-msg=C0103

# ``import calypso`` refers to the ``calypso`` module, not ``calypso.py`` 
# pylint: disable-msg=W0406

"""
Calypso Server entry point.

Launch the Calypso Server according to configuration and command-line
arguments.

"""

# TODO: Manage smart and configurable logs

import logging
import os
import sys
import optparse

import calypso
import calypso.webdav as webdav

# Get command-line options
parser = optparse.OptionParser()
parser.add_option(
    "-v", "--version", action="store_true",
    default=False,
    help="show version and exit")
parser.add_option(
    "-d", "--daemon", action="store_true",
    default=calypso.config.getboolean("server", "daemon"),
    help="launch as daemon")
parser.add_option(
    "-f", "--foreground", action="store_false", dest="daemon",
    help="launch in foreground (opposite of --daemon)")
parser.add_option(
    "-H", "--host",
    default=calypso.config.get("server", "host"),
    help="set server hostname")
parser.add_option(
    "-p", "--port", type="int",
    default=calypso.config.getint("server", "port"),
    help="set server port")
parser.add_option(
    "-s", "--ssl", action="store_true",
    default=calypso.config.getboolean("server", "ssl"),
    help="use SSL connection")
parser.add_option(
    "-S", "--no-ssl", action="store_false", dest="ssl",
    help="do not use SSL connection (opposite of --ssl)")
parser.add_option(
    "-k", "--key",
    default=calypso.config.get("server", "key"),
    help="private key file ")
parser.add_option(
    "-c", "--certificate",
    default=calypso.config.get("server", "certificate"),
    help="certificate file ")
parser.add_option(
    "-i", "--import", dest="import_dest")
parser.add_option(
    "-g", "--debug", action="store_true",
    default=False,
    help="enable debug logging")
    
(options, args) = parser.parse_args()

# Update Calypso configuration according to options
for option in parser.option_list:
    key = option.dest
    if key:
        value = getattr(options, key)
        calypso.config.set("server", key, value)

# Print version and exit if the option is given
if options.version:
    print(calypso.VERSION)
    sys.exit()

log = logging.getLogger()
ch = logging.StreamHandler()

# Handle debugging option and log levels
if options.debug:
    log.setLevel(logging.DEBUG)
    ch.setLevel(logging.DEBUG)
    logging.basicConfig(level=logging.DEBUG)
    log.debug("enable debugging")
else:
    log.setLevel(logging.WARN)
    ch.setLevel(logging.WARN)
    logging.basicConfig(level=logging.WARN)
    

# Run import if requested
if options.import_dest:
    try:
        collection = webdav.Collection(options.import_dest)
    except Exception:
        print "Cannot open collection %s" % options.import_dest
        sys.exit(1)
    success = True
    for arg in args:
        if not collection.import_file(arg):
            success = False
    if success:
        sys.exit(0)
    else:
        sys.exit(1)

# Fork if Calypso is launched as daemon
if options.daemon:
    if os.fork():
        sys.exit()
    sys.stdout = sys.stderr = open(os.devnull, "w")

# Launch server
server_class = calypso.HTTPSServer if options.ssl else calypso.HTTPServer
server = server_class(
    (options.host, options.port), calypso.CollectionHTTPHandler)
server.serve_forever(poll_interval=10)
